components/plan-build-ps1/bin/hab-plan-build.ps1

```diff
@@ -1532,51 +1532,43 @@ if (-Not (Test-Path "$Context")) {
-#   `./habitat/$pkg_target/plan.ps1`
-# In most cases, Plan authors should use the default location of `./plan.ps1`
-# or `./habitat/plan.ps1`.  The exception to this is when the $pkg_target
-# requires variations to the default `plan.ps1`. Plan authors can create these
-# variants by placing a plan file in the appropriate $pkg_target directory
-# relative to the default plan.ps1.
-#
-# With plan variants, plans can exist in 4 places per $pkg_target relative to
-# the `$PLAN_CONTEXT` directory. Only two combinations are allowed:
-#
-#   `./plan.ps1` AND `./$pkg_target/plan.ps1`
-#   OR
-#   `./habitat/plan.ps1` AND `./habitat/$pkg_target/plan.ps1`
-# Consider all other combination of two plans invalid and abort if found.
-
-# ** Internal ** Relative to the current plan context,  check for a variant
-#   that matches the current $pkg_target, and update $PLAN_CONTEXT if found.
-function Set-PlanContextFromTarget {
-    if (Test-Path "$PLAN_CONTEXT\$pkg_target\plan.ps1") {
-        Write-BuildLine "Detected Plan Variant!"
-        $script:PLAN_CONTEXT = "$PLAN_CONTEXT\$pkg_target"
-        Write-BuildLine "$PLAN_CONTEXT"
-    }
-}
-
-# Look for a plan.ps1 relative to the $PLAN_CONTEXT. If we find an invalid
-#   combination or are unable to find a plan.ps1,  abort with a message to the
-#   user with the failure case.
-if (Test-Path "$PLAN_CONTEXT\plan.ps1") {
-    if (Test-Path "$PLAN_CONTEXT\habitat\plan.ps1") {
-        $places = "$PLAN_CONTEXT\plan.ps1 and $PLAN_CONTEXT\habitat\plan.ps1"
-        throw "A Plan file was found at $places. Only one is allowed at a time"
-    }
-    Set-PlanContextFromTarget
-} elseif (Test-Path "$PLAN_CONTEXT\habitat\plan.ps1") {
-    $script:PLAN_CONTEXT = "$PLAN_CONTEXT\habitat"
-    Set-PlanContextFromTarget
-} else {
-    $places = "$PLAN_CONTEXT\plan.ps1 or $PLAN_CONTEXT\habitat\plan.ps1"
-    throw "Plan file not found at $places"
+
+# Look for a plan.ps1 relative to the $PLAN_CONTEXT. Acceptable locations are:
+#   "$PLAN_CONTEXT\plan.ps1",
+#   "$PLAN_CONTEXT\habitat\plan.ps1",
+#   "$PLAN_CONTEXT\$pkg_target\plan.ps1",
+#   "$PLAN_CONTEXT\habitat\$pkg_target\plan.ps1"
+# A plan found in the target folder will take presedence above a non-target
+# folder. If we find an invalid combination or are unable to find a plan.ps1,
+# abort with a message to the user with the failure case.
+$paths = @()
+$candidatePaths = @(
+    "$PLAN_CONTEXT\plan.ps1",
+    "$PLAN_CONTEXT\habitat\plan.ps1"
+)
+$candidateTargetPaths = @(
+    "$PLAN_CONTEXT\$pkg_target\plan.ps1",
+    "$PLAN_CONTEXT\habitat\$pkg_target\plan.ps1"
+)
+foreach($place in $candidateTargetPaths) {
+    if(Test-Path $place) {
+        $paths += $place
+        $script:PLAN_CONTEXT = (Split-Path $place -Parent)
+    }
+}
+if($paths.Count -eq 0) {
+    foreach($place in $candidatePaths) {
+        if(Test-Path $place) {
+            $paths += $place
+            $script:PLAN_CONTEXT = (Split-Path $place -Parent)
+        }
+    }
+}
```

Comment:
I'll add some clarifying comments to the code but the answer is that we want to be able to see if the user has more than one candidate plan later triggering the `if($paths.Count -gt 1)` error condition. We do need to track the target and non-target candidates separately because it is OK to have a target plan and a non-target candidate plan. In such a case the target plan takes precedence.

Reply:
I don't think this logic works as described.

Here's the overall order in which things are checked (targets first, then "plain" plans):
- "$PLAN_CONTEXT\$pkg_target\plan.ps1",
- "$PLAN_CONTEXT\habitat\$pkg_target\plan.ps1"
- "$PLAN_CONTEXT\plan.ps1",
- "$PLAN_CONTEXT\habitat\plan.ps1"

If you have a plan at `$pkg_target/plan.ps1` and `plan.ps1`, the logic below will claim to use `$pkg_target/plan.ps1`, but the `$script:PLAN_CONTEXT` variable would have been set for the `plan.ps1` plan. This is because the value of `$script:PLAN_CONTEXT` is set to the last existing path that we check, regardless of what came before.

I understand the need to track targeted plans separately from non-targeted plans, but I think this could be simplified by gathering just the path information first, checking to see if you've got the right number (issuing any relevant warnings in the process), and then once you are certain you have one path to operate on, determine what the value of `$script:PLAN_CONTEXT` is. In other words, only do that once, rather than multiple times in a loop. That decouples the two logical operations that we are doing and makes it much more clear what is going on.

The same concerns apply for the Bash implementation.

Reply:
I'm also wondering if using the variable name `$PLAN_CONTEXT` here is appropriate. We're actually trying to figure out what the ultimate `$PLAN_CONTEXT` is, right? So `$PLAN_CONTEXT` at _this_ point is really just the initial or root directory that we're searching in, correct? We don't know what the plan is, so we _can't_ know what that plan's context is.

Not overloading variable names (particularly in this corner of the code :sweat_smile:) is going to make it much easier to maintain this code.

Reply:
> If you have a plan at $pkg_target/plan.ps1 and plan.ps1, the logic below will claim to use $pkg_target/plan.ps1, but the $script:PLAN_CONTEXT variable would have been set for the plan.ps1 plan. This is because the value of $script:PLAN_CONTEXT is set to the last existing path that we check, regardless of what came before.

Son of a...@$$$#!